### PR TITLE
add policy msbuild.multi_tool_task

### PR DESCRIPTION
**Pull request by ChrisCatCP**

### CODE REVIEW

**Comment by SirLynix** [comment]
if its a msbuild config, setting it as a toolchain runenvs won't do anything, as xmake doesn't use msbuild.exe (it uses cl.exe and link.exe with its own scheduler).

**Comment by ChrisCatCP** [comment]
> if its a msbuild config, setting it as a toolchain runenvs won't do anything, as xmake doesn't use msbuild.exe (it uses cl.exe and link.exe with its own scheduler).

it used by package(cmake or msbuild)

**Comment by SirLynix** [comment]
then maybe the policy shouldn't be named that way. since it's just an env variable, what's the benefit of adding it to xmake as a policy instead of just setting the env variable on the computer using xmake?

**Comment by ChrisCatCP** [comment]
> then maybe the policy shouldn't be named that way. since it's just an env variable, what's the benefit of adding it to xmake as a policy instead of just setting the env variable on the computer using xmake?

Just to speed up package compile  like package.cmake_generator.ninja

**Comment by SirLynix** [comment]
As far as I know ninja doesn't use msbuild either, that's the purpose of ninja (to schedule compiler calls like xmake does), but I may be wrong.

however, does it necessary need to be a xmake policy if you want set the environment variable in your xmake.lua, or on the CI/computer using it? there's already a lot of policies

**Comment by ChrisCatCP** [comment]
If package.cmake_gener.ninja is not enabled, this may be useful on msvc

**Comment by SirLynix** [comment]
you're not answering my question, what's the point of having it in xmake instead of just using `os.setenv("UseMultiToolTask", "true")` in your xmake.lua, or set `UseMultiToolTask` as an environment variable on the computer compiling?

**Comment by ChrisCatCP** [comment]
> you're not answering my question, what's the point of having it in xmake instead of just using `os.setenv("UseMultiToolTask", "true")` in your xmake.lua, or set `UseMultiToolTask` as an environment variable on the computer compiling?

Is os.setenv ("CMAKE_GENERATOR", "Ninja") and package.cmake_gener.ninja similar to this?

**Comment by star-hengxing** [comment]
I think it should add here without policy
https://github.com/xmake-io/xmake/blob/2412ed94e70b00b0c2f3bd8c4553828efd458b1b/xmake/modules/package/tools/msbuild.lua#L60

**Comment by ChrisCatCP** [comment]
> I think it should add here
> 
> https://github.com/xmake-io/xmake/blob/2412ed94e70b00b0c2f3bd8c4553828efd458b1b/xmake/modules/package/tools/msbuild.lua#L60

cmake dont use package.tools.msbuild

**Comment by SirLynix** [comment]
> > you're not answering my question, what's the point of having it in xmake instead of just using `os.setenv("UseMultiToolTask", "true")` in your xmake.lua, or set `UseMultiToolTask` as an environment variable on the computer compiling?
> 
> Is os.setenv ("CMAKE_GENERATOR", "Ninja") and package.cmake_gener.ninja similar to this?

it's not really the same as the env is taken by xmake directly, and xmake has full control on the way cmake is invoked

**Comment by ChrisCatCP** [comment]
xmake also has full control  invoke msbuild，Do you think msbuild /p:UseMultiToolTask=true would be better?

**Comment by waruqi** [comment]
既然仅仅对 package 里面 tools.cmake 生效，添加到 msvc toolchain 干嘛，哪里是调用 cl.exe 构建用的，即使也会影响到 tools.cmake ，但是影响范围太广，会干扰到原本不应该被影响的地方。。

应该只在 tools.cmake 去修改，而且 msbuild 并行，这里应该已经加了，不需要再加额外的配置，`msbuild -m:jobs` 已经指定了并行，除非它没生效

https://github.com/xmake-io/xmake/blob/c6a1199e622bf6de11a8ab0cd7665a39d8fd1c36/xmake/modules/package/tools/cmake.lua#L1021

即使 `/m:` 不生效，必须要加额外的 envs，也应该 https://github.com/xmake-io/xmake/blob/c6a1199e622bf6de11a8ab0cd7665a39d8fd1c36/xmake/modules/package/tools/cmake.lua#L96 这里加。。

另外，先得确认为啥 `/m:` 为啥不管用，有验证过么

**Comment by ChrisCatCP** [comment]
> 既然仅仅对 package 里面 tools.cmake 生效，添加到 msvc toolchain 干嘛，哪里是调用 cl.exe 构建用的，即使也会影响到 tools.cmake ，但是影响范围太广，会干扰到原本不应该被影响的地方。。
> 
> 应该只在 tools.cmake 去修改，而且 msbuild 并行，这里应该已经加了，不需要再加额外的配置，`msbuild -m:jobs` 已经指定了并行，除非它没生效
> 
> https://github.com/xmake-io/xmake/blob/c6a1199e622bf6de11a8ab0cd7665a39d8fd1c36/xmake/modules/package/tools/cmake.lua#L1021
> 
> 即使 `/m:` 不生效，必须要加额外的 envs，也应该
> 
> https://github.com/xmake-io/xmake/blob/c6a1199e622bf6de11a8ab0cd7665a39d8fd1c36/xmake/modules/package/tools/cmake.lua#L96
> 
> 这里加。。
> 另外，先得确认为啥 `/m:` 为啥不管用，有验证过么

/m:不是不生效，/m是针对一个sln的多个项目并行编译，这个是针对源文件进行并行编译，而且不仅仅针对cmake生效，也针对package.tool.msbuild生效，所以加到toolchain里了

**Comment by waruqi** [comment]
> /m:不是不生效，/m是针对一个sln的多个项目并行编译，这个是针对源文件进行并行编译，而且不仅仅针对cmake生效，也针对package.tool.msbuild生效，所以加到toolchain里了

那直接在 `tools.cmake` 和 `tools.msbuild` 的 envs 里加上不就好了么，哪里需要生效，就加哪里。。

加 toolchain，可不只是 tools 里生效这么简单的，这个影响的范围大了去了。。会在大量无意义的地方，添加上这个 env 。。

**Comment by ChrisCatCP** [comment]
> > /m:不是不生效，/m是针对一个sln的多个项目并行编译，这个是针对源文件进行并行编译，而且不仅仅针对cmake生效，也针对package.tool.msbuild生效，所以加到toolchain里了
> 
> 那直接在 `tools.cmake` 和 `tools.msbuild` 的 envs 里加上不就好了么，哪里需要生效，就加哪里。。
> 
> 加 toolchain，可不只是 tools 里生效这么简单的，这个影响的范围大了去了。。会在大量无意义的地方，添加上这个 env 。。

是这样没错，但是这样改动少，而且这个环境变量只影响msbuild，不会影响别的工具链，如果要加到特定工具里，那就直接加到msbuild的参数里可能更好一点  /p:UseMultiToolTask=true

**Comment by waruqi** [comment]
> > > /m:不是不生效，/m是针对一个sln的多个项目并行编译，这个是针对源文件进行并行编译，而且不仅仅针对cmake生效，也针对package.tool.msbuild生效，所以加到toolchain里了
> > 
> > 
> > 那直接在 `tools.cmake` 和 `tools.msbuild` 的 envs 里加上不就好了么，哪里需要生效，就加哪里。。
> > 加 toolchain，可不只是 tools 里生效这么简单的，这个影响的范围大了去了。。会在大量无意义的地方，添加上这个 env 。。
> 
> 是这样没错，但是这样改动少，而且这个环境变量只影响msbuild，不会影响别的工具链，如果要加到特定工具里，那就直接加到msbuild的参数里可能更好一点 /p:UseMultiToolTask=true

你这样改动反而最多。。clang-cl 也需要， package 切换 clang-cl ，msbuild 也会调用到，那么 clang-cl toolchain 的也得同步改。。而且 msvc/clang-cl toolchain 本身跟 msbuild 没有必然关系。。你去加 toolchain 。。属于反向操作。。没这么改法的。。其他开发者会看的莫名其妙。我都不保证 toolchain 一定会被 msbuild 用到，也许将来我重构下，msbuild 不会去调用 toolchain 的 envs 呢，那不就被 break 了么。。但是你改 `tools/cmake`, `tools/msbuild` 肯定会一直生效，并且其他维护者看到 也能一目了然

`tools.cmake` 和 `tools.msbuild` 也就改两个地方，改动量又不大。。

反正，我还是那句话，哪里实际需要，就改哪里，不要改任何其他无意义的地方。

**Comment by ChrisCatCP** [comment]
> > > > /m:不是不生效，/m是针对一个sln的多个项目并行编译，这个是针对源文件进行并行编译，而且不仅仅针对cmake生效，也针对package.tool.msbuild生效，所以加到toolchain里了
> > > 
> > > 
> > > 那直接在 `tools.cmake` 和 `tools.msbuild` 的 envs 里加上不就好了么，哪里需要生效，就加哪里。。
> > > 加 toolchain，可不只是 tools 里生效这么简单的，这个影响的范围大了去了。。会在大量无意义的地方，添加上这个 env 。。
> > 
> > 
> > 是这样没错，但是这样改动少，而且这个环境变量只影响msbuild，不会影响别的工具链，如果要加到特定工具里，那就直接加到msbuild的参数里可能更好一点 /p:UseMultiToolTask=true
> 
> 你这样改动反而最多。。clang-cl 也需要， package 切换 clang-cl ，msbuild 也会调用到，那么 clang-cl toolchain 的也得同步改。。而且 msvc/clang-cl toolchain 本身跟 msbuild 没有必然关系。。你去加 toolchain 。。属于反向操作。。没这么改法的。。其他开发者会看的莫名其妙。我都不保证 toolchain 一定会被 msbuild 用到，也许将来我重构下，msbuild 不会去调用 toolchain 的 envs 呢，那不就被 break 了么。。但是你改 `tools/cmake`, `tools/msbuild` 肯定会一直生效，并且其他维护者看到 也能一目了然
> 
> `tools.cmake` 和 `tools.msbuild` 也就改两个地方，改动量又不大。。
> 
> 反正，我还是那句话，哪里实际需要，就改哪里，不要改任何其他无意义的地方。

已修改

File: xmake/modules/package/tools/cmake.lua [hunk 1]
```diff
@@ -945,14 +946,8 @@ end
...
 -- do build for msvc
 function _build_for_msvc(package, configs, opt)
-    local jobs = _get_parallel_njobs(opt)
     local slnfile = assert(find_file("*.sln", os.curdir()), "*.sln file not found!")
-    local runenvs = _get_msvc_runenvs(package)
-    local msbuild = find_tool("msbuild", {envs = runenvs})
-    os.vrunv(msbuild.program, {slnfile, "-nologo", "-t:Rebuild",
-            (jobs ~= nil and format("-m:%d", jobs) or "-m"),
-            "-p:Configuration=" .. (package:is_debug() and "Debug" or "Release"),
-            "-p:Platform=" .. _get_vsarch(package)}, {envs = runenvs})
```
Comment (waruqi) [inline xmake/modules/package/tools/cmake.lua:955]: 如果 cmake.lua 里没再用到 _get_vsarch，可以删了

Comment (ChrisCatCP) [reply xmake/modules/package/tools/cmake.lua:955]: 已修改

**Pull request merged**